### PR TITLE
[llvm] use 64-bit types for result of getDwarfRegNum (NFC)

### PR DESCRIPTION
**Pull request by willghatch**
The register encoding used by NVPTX and cuda-gdb basically use strings encoded as numbers.  They are always within 64-bits, but typically outside of 32-bits, since they often need at least 5 characters.

This patch changes the signature of `MCRegisterInfo::getDwarfRegNum` and some related data structures to use 64-bit numbers to accommodate encodings like this.

Additionally, `MCRegisterInfo::getDwarfRegNum` is marked as virtual, so that targets with peculiar dwarf register mapping schemes (such as NVPTX) can override its behavior.

I originally tried to do a broader switch to 64-bit types for registers, but it caused many problems. There are various places in code generation where registers are not just treated as 32-bit numbers, but also treat certain bit offsets as flags. So I limited the change as much as possible to just the output of `getDwarfRegNum`. Keeping the types used by `DwarfLLVMRegPair` as unsigned preserves the current behaviors. The only way to give a 64-bit output from `getDwarfRegNum` that actually needs more than 32-bits is to override `getDwarfRegNum` and provide an implementation that sidesteps the use of the `DwarfLLVMRegPair` maps defined in tablegen files.

First layer of stack supporting: https://github.com/llvm/llvm-project/pull/109495

### CODE REVIEW

File: llvm/lib/MC/MCRegisterInfo.cpp [hunk 1]
```diff
@@ -151,24 +151,28 @@ int MCRegisterInfo::getDwarfRegNum(MCRegister RegNum, bool isEH) const {
   const DwarfLLVMRegPair *I = std::lower_bound(M, M+Size, Key);
   if (I == M+Size || I->FromReg != RegNum)
     return -1;
-  return I->ToReg;
+  // Consumers need to be able to detect -1 and -2, but at various points
+  // the numbers move between unsigned and signed representations, as well as
+  // between 32- and 64-bit representations. We need to convert first to int
+  // before int64_t for proper sign handling.
+  return int64_t(int(I->ToReg));
```
Comment (walter-erquinigo) [inline llvm/lib/MC/MCRegisterInfo.cpp:158]: what about doing the following:
```
auto reg = I->ToReg;
switch (reg) {
  case -1: return -1;
  case -2: return -2;
  default: return reg;
}
```

File: llvm/lib/Target/Lanai/LanaiRegisterInfo.h [hunk 2]
```diff
@@ -43,8 +43,6 @@ struct LanaiRegisterInfo : public LanaiGenRegisterInfo {
   Register getFrameRegister(const MachineFunction &MF) const override;
   Register getBaseRegister() const;
   bool hasBasePointer(const MachineFunction &MF) const;
-
-  int getDwarfRegNum(unsigned RegNum, bool IsEH) const;
```
Comment (walter-erquinigo) [inline llvm/lib/Target/Lanai/LanaiRegisterInfo.h:47]: why did you remove this?

File: llvm/lib/MC/MCRegisterInfo.cpp [hunk 3]
```diff
@@ -151,24 +151,28 @@ int MCRegisterInfo::getDwarfRegNum(MCRegister RegNum, bool isEH) const {
...
-std::optional<MCRegister> MCRegisterInfo::getLLVMRegNum(unsigned RegNum,
+std::optional<MCRegister> MCRegisterInfo::getLLVMRegNum(uint64_t RegNum,
                                                         bool isEH) const {
   const DwarfLLVMRegPair *M = isEH ? EHDwarf2LRegs : Dwarf2LRegs;
   unsigned Size = isEH ? EHDwarf2LRegsSize : Dwarf2LRegsSize;
 
   if (!M)
     return std::nullopt;
-  DwarfLLVMRegPair Key = { RegNum, 0 };
+  DwarfLLVMRegPair Key = {unsigned(RegNum), 0};
```
Comment (walter-erquinigo) [inline llvm/lib/MC/MCRegisterInfo.cpp:168]: wouldn't it be better to have `DwarfLLVMRegPair` support int64_t? Or is there something that makes you sure that the current conversion is safe?

File: llvm/lib/Target/Lanai/LanaiRegisterInfo.h (hunk 2 above)
Comment (willghatch) [reply llvm/lib/Target/Lanai/LanaiRegisterInfo.h:47]: It caused errors when I switched `getDwarfRegNum` to be virtual, and there didn't seem to be a reason for it to be there.  No tests failed from removing it.  I believe it is vestigial, but I'm open to someone telling or showing me why it should be there.

File: llvm/lib/MC/MCRegisterInfo.cpp (hunk 3 above)
Comment (willghatch) [reply llvm/lib/MC/MCRegisterInfo.cpp:168]: I originally tried to do a broader switch to 64-bit types for registers, but it caused many problems.  There are various places in code generation where registers are not just treated as 32-bit numbers, but also treat certain bit offsets as flags.  So I limited the change as much as possible to just the output of `getDwarfRegNum`.  Keeping the types used by `DwarfLLVMRegPair` as `unsigned` preserves the current behaviors.  The only way to give a 64-bit output from `getDwarfRegNum` that actually needs more than 32-bits is to override `getDwarfRegNum` and provide an implementation that sidesteps these maps.  These maps are generated from tablegen register definitions, and don't make sense for an implementation that needs to use a weird encoding that requires 64 bits anyway.

File: llvm/lib/MC/MCRegisterInfo.cpp (hunk 1 above)
Comment (willghatch) [reply llvm/lib/MC/MCRegisterInfo.cpp:158]: That might be good.  I'm not certain there are not other negative values currently being used as signal values, but if that switch passes the tests then it's probably fine.  I'm not certain that's necessarily better, though.

File: llvm/lib/Target/Lanai/LanaiRegisterInfo.h (hunk 2 above)
Comment (walter-erquinigo) [reply llvm/lib/Target/Lanai/LanaiRegisterInfo.h:47]: No worries. What you say makes sense.

File: llvm/lib/MC/MCRegisterInfo.cpp (hunk 3 above)
Comment (walter-erquinigo) [reply llvm/lib/MC/MCRegisterInfo.cpp:168]: Oof. Got it. Please mention this in the commit description for posterity.

File: llvm/lib/MC/MCRegisterInfo.cpp (hunk 1 above)
Comment (walter-erquinigo) [reply llvm/lib/MC/MCRegisterInfo.cpp:158]: Then, what about `int64_t(int32_t(I->ToReg))`?

Comment (willghatch) [reply llvm/lib/MC/MCRegisterInfo.cpp:158]: All of the existing code dealing with these register numbers uses the types `int` and `unsigned`.  The `int64_t(int(I->ToReg))` conversion seems more consistent with other usage than `int64_t(int32_t(I->ToReg))`.

**Review by walter-erquinigo — APPROVED** [review]
